exercises/edward_brown/ex01_basics/include/MakeWordFrequencyMap.h

```diff
@@ -0,0 +1,45 @@
+#ifndef EX1_MAKE_WORD_FREQUENCY_MAP
+#define EX1_MAKE_WORD_FREQUENCY_MAP
+#include "ForEachWord.h"
+#include "Punctuation.h"
+#include "ToLowerCase.h"
+#include <unordered_map>
+
+/** A predicate which determines if CandidateWord is a word for the purposes of
+ *  word frequency analysis.
+ *
+ *  @param CandidateWord The candidate to check.
+ *  @return True if CandidateWord is a 'word' false if not.
+ */
+bool isWord(std::string const &CandidateWord) {
+  auto constexpr MINIMUM_WORD_LENGTH = 4;
+  return CandidateWord.size() > MINIMUM_WORD_LENGTH;
+}
+
+/** Creates a map from strings containing words found in the range [Begin, End)
+ * to integer values denoting the corresponding number of occurences in said
+ * range.
+ *
+ *  This analysis is case and punctuation insensitive. Words with a length
+ * shorter than four characters are ignored.
+ *
+ *  @param Begin The const_iterator marking the beginning of the range to search
+ * for words.
+ *  @param End The const_iterator marking the end of the range to search for
+ * words.
+ *  @return The map built from the word frequency analysis performed.
+ */
+template <typename ConstCharInputIterator>
+std::unordered_map<std::string, int>
+makeWordFrequencyMap(ConstCharInputIterator begin, ConstCharInputIterator end) {
+  auto wordFrequencyMap = std::unordered_map<std::string, int>();
+  forEachWord(
+      begin, end, [&wordFrequencyMap](std::string WordCandidate) -> void {
+        WordCandidate = stripPunctuation(WordCandidate);
+        if (isWord(WordCandidate)) {
+          wordFrequencyMap[toLowerCase(stripPunctuation(WordCandidate))]++;
+        }
+      });
+  return std::move(wordFrequencyMap);
```

Comment:
After some research it appears it's not a good idea. I have recieved compiler warnings about unecessary `move`s on return values in the past when compiling c++14 code with `-Wall` set but didn't remember getting anything with a c++11 compiler so assumed that the copy elision rules were somehow enhanced in c++14. 

However the following paragraph from [here](http://en.cppreference.com/w/cpp/language/copy_elision) reveals otherwise. 

> If a function returns a class type by value, and the return statement's expression is the name of a non-volatile object with automatic storage duration, which isn't the function parameter, or a catch clause parameter, and which has the same type (ignoring top-level cv-qualification) as the return type of the function, then copy/move (since C++11) is omitted. When that local object is constructed, it is constructed directly in the storage where the function's return value would otherwise be moved or copied to. This variant of copy elision is known as NRVO, "named return value optimization"

I have removed the extra move in 37d7595635f1bb589548e3e2f91bfef100e2c391